examples/config.c

```diff
@@ -0,0 +1,304 @@
+
+	bool has_width = false;
+	bool has_height = false;
+	bool has_x = false;
+	bool has_y = false;
+
+	char *pch = strtok(buf, "x+");
+	while (pch != NULL) {
+		errno = 0;
+		char *endptr;
+		long val = strtol(pch, &endptr, 0);
+
+		if ((errno == ERANGE && (val == LONG_MAX || val == LONG_MIN))
+				|| (errno != 0 && val == 0)) {
+			goto invalid_input;
+		}
+
+		if (endptr == pch) {
+			goto invalid_input;
+		}
+
+		if (!has_width) {
+			geo->width = val;
+			has_width = true;
+		} else if (!has_height) {
+			geo->height = val;
+			has_height = true;
+		} else if (!has_x) {
+			geo->x = val;
+			has_x = true;
+		} else if (!has_y) {
+			geo->y = val;
+			has_y = true;
+		} else {
+			break;
+		}
+		pch = strtok(NULL, "x+");
+	}
+
+	if (!has_width || !has_height || !has_x || !has_y) {
+		goto invalid_input;
+	}
+
+	free(buf);
+	return geo;
+
+invalid_input:
+	wlr_log(L_ERROR, "could not parse geometry string: %s", str);
+	free(buf);
+	free(geo);
+	return NULL;
+}
+
+static const char *output_prefix = "output:";
+static const char *device_prefix = "device:";
+
+static int config_ini_handler(void *user, const char *section, const char *name, const char *value) {
+	struct example_config *config = user;
+	if (strncmp(output_prefix, section, strlen(output_prefix)) == 0) {
+		const char *output_name = section + strlen(output_prefix);
+		struct output_config *oc;
+		bool found = false;
+
+		wl_list_for_each(oc, &config->outputs, link) {
+			if (strcmp(oc->name, output_name) == 0) {
+				found = true;
+				break;
+			}
+		}
+
+		if (!found) {
+			oc = calloc(1, sizeof(struct output_config));
+			oc->name = strdup(output_name);
+			oc->transform = WL_OUTPUT_TRANSFORM_NORMAL;
+			wl_list_insert(&config->outputs, &oc->link);
+		}
+
+		if (strcmp(name, "x") == 0) {
+			oc->x = strtol(value, NULL, 10);
+		} else if (strcmp(name, "y") == 0) {
+			oc->y = strtol(value, NULL, 10);
+		} else if (strcmp(name, "rotate") == 0) {
+			if (strcmp(value, "90") == 0) {
+				oc->transform = WL_OUTPUT_TRANSFORM_90;
+			} else if (strcmp(value, "180") == 0) {
+				oc->transform = WL_OUTPUT_TRANSFORM_180;
+			} else if (strcmp(value, "270") == 0) {
+				oc->transform = WL_OUTPUT_TRANSFORM_270;
+			} else if (strcmp(value, "flipped") == 0) {
+				oc->transform = WL_OUTPUT_TRANSFORM_FLIPPED;
+			} else if (strcmp(value, "flipped-90") == 0) {
+				oc->transform = WL_OUTPUT_TRANSFORM_FLIPPED_90;
+			} else if (strcmp(value, "flipped-180") == 0) {
+				oc->transform = WL_OUTPUT_TRANSFORM_FLIPPED_180;
+			} else if (strcmp(value, "flipped-270") == 0) {
+				oc->transform = WL_OUTPUT_TRANSFORM_FLIPPED_270;
+			} else {
+				wlr_log(L_ERROR, "got unknown transform value: %s", value);
+			}
+		}
+	} else if (strcmp(section, "cursor") == 0) {
+		if (strcmp(name, "map-to-output") == 0) {
+			config->cursor.mapped_output = strdup(value);
+		} else if (strcmp(name, "geometry") == 0) {
+			if (config->cursor.mapped_geo) {
+				free(config->cursor.mapped_geo);
```

Comment:
(nitpick) feels weird to have this one freed on multi-value but not others (e.g. mapped_output just above)
(also, free can be called on NULL values, not sure if we have a policy about checking for null before free)